languages/python/sqlalchemy-oso/sqlalchemy_oso/roles2.py

```diff
@@ -175,14 +197,303 @@ class Role:
 @dataclass
 class Resource:
     name: str
+    type: str
     python_class: Any
     actions: Set[str]
     roles: Set[str]
 
 
+@dataclass
+class Config:
+    resources: Dict[str, Resource]
+    permissions: List[Permission]
+    roles: Dict[str, Role]
+    relationships: List[Relationship]
+
+
+def parse_permission(permission, python_class, config):
+    """ Parse a permission string, check if it's valid and return a Permission """
+    if ":" in permission:
+        resource_name, action = permission.split(":", 1)
+        if resource_name not in config.resources:
+            raise OsoError("Invalid permission namespace.")
+        permission_python_class = config.resources[resource_name].python_class
+    else:
+        action = permission
+        permission_python_class = python_class
+    perm = Permission(
+        name=action,
+        type=permission_python_class.__name__,
+        python_class=permission_python_class,
+    )
+    if perm not in config.permissions:
+        raise OsoError(
+            f"Permission {perm.name} doesn't exist for resource {perm.type}."
+        )
+    return perm
+
+
+def read_config(oso):
+    """
+    Queries the polar policy for the configuration
+    """
+    config = Config(resources={}, permissions=[], roles={}, relationships=[])
+
+    # Register relationships
+    role_relationships = oso.query_rule(
+        "parent",
+        Variable("resource"),
+        Variable("parent_resource"),
+        accept_expression=True,
+    )
+
+    # Currently there is only one valid relationship, a parent.
+    # There is also only one way you can write it as a rule in polar.
+    # parent(child_resource, parent_resource) if
+    #     child.parent = parent_resource;
+    #
+    # @TODO: Support other forms of this rule, eg
+    # parent(child_resource, parent_resource) if
+    #     child.parent_id = parent_resource.id;
+    for result in role_relationships:
+        try:
+            constraints = result["bindings"]["resource"]
+            assert len(constraints.args) == 2
+            type_check = constraints.args[0]
+            child_t = isa_type(type_check)
+            get_parent = constraints.args[1]
+            assert get_parent.operator == "Isa"
+            assert len(get_parent.args) == 2
+            getter = get_parent.args[0]
+            assert getter.operator == "Dot"
+            assert len(getter.args) == 2
+            assert getter.args[0] == Variable("_this")
+            child_attr = getter.args[1]
+            pattern = get_parent.args[1]
+            parent_t = pattern.tag
```

Comment:
Nit, but can we spell out `tag` instead of just using `t` throughout?

Reply:
will spell out type